File: Java/src/main/java/PowerOfTwo.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

public class PowerOfTwo {
    private static PrintWriter logFile;
    
    static {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            if (logFile != null) {
                logFile.flush();
                logFile.close();
            }
        }));
    }
    
    public static void main(String[] args) {
        String logPath = getLogPath();
        
        try {
            logFile = new PrintWriter(new FileWriter(logPath));
        } catch (IOException e) {
            System.out.println("Ошибка: не удалось открыть файл для записи");
            return;
        }
        
        logFile.println("Вычисление степеней двойки (Big Integer)");
        logFile.println("========================================\n");
        logFile.flush();
        
        System.out.println("Вычисление степеней двойки (Big Integer)");
        System.out.println("========================================\n");
        
        BigInt value = new BigInt(1);
        
        for (long power = 0; ; power++) {
            String result = String.format("2^%d = %s", power, value.toString());
            logFile.println(result);
            logFile.flush();
            System.out.println(result);
            
            value.mul2();
        }
    }
    
    private static String getLogPath() {
        try {
            String currentDir = System.getProperty("user.dir");
            return currentDir + File.separator + "PowerOfTwo.log";
        } catch (Exception e) {
            return "PowerOfTwo.log";
        }
    }
}

File: Pascal/PowerOfTwo.pas
{$mode objfpc}{$H+}

program PowerOfTwo;

uses
  SysUtils, Classes;

type
  TBigInt = class
  private
    digits: array of Integer;
    count: Integer;
  public
    constructor Create(value: Integer);
    destructor Destroy; override;
    procedure MultiplyBy2;
    function ToString: string; override;
  end;

constructor TBigInt.Create(value: Integer);
begin
  inherited Create;
  SetLength(digits, 1);
  digits[0] := value;
  count := 1;
end;

destructor TBigInt.Destroy;
begin
  SetLength(digits, 0);
  inherited;
end;

procedure TBigInt.MultiplyBy2;
var
  i, product, carry: Integer;
begin
  carry := 0;
  for i := 0 to count - 1 do
  begin
    product := digits[i] * 2 + carry;
    digits[i] := product mod 10;
    carry := product div 10;
  end;

  while carry > 0 do
  begin
    if count >= Length(digits) then
      SetLength(digits, count + 10);
    digits[count] := carry mod 10;
    carry := carry div 10;
    Inc(count);
  end;
end;

function TBigInt.ToString: string;
var
  i: Integer;
begin
  Result := '';
  for i := count - 1 downto 0 do
    Result := Result + IntToStr(digits[i]);
end;

var
  logFile: TextFile;
  logPath: string;
  value: TBigInt;
  power: Int64;
  result: string;
  header: string;
  separator: string;

begin
  { Get log file path }
  logPath := GetCurrentDir + PathDelim + 'PowerOfTwo.log';

  { Create/overwrite log file }
  AssignFile(logFile, logPath);
  Rewrite(logFile);

  header := 'Computing powers of two (Big Integer)';
  separator := '======================================';

  WriteLn(logFile, header);
  WriteLn(logFile, separator);
  WriteLn(logFile, '');

  WriteLn(header);
  WriteLn(separator);
  WriteLn('');

  value := TBigInt.Create(1);
  power := 0;

  try
    while True do
    begin
      result := '2^' + IntToStr(power) + ' = ' + value.ToString;

      WriteLn(logFile, result);
      Flush(logFile);
      WriteLn(result);

      value.MultiplyBy2;
      Inc(power);
    end;
  finally
    value.Free;
    CloseFile(logFile);
  end;
end.
